### PR TITLE
storage - NBDE playground 

### PR DESCRIPTION
**Pull request by mvollmer**
The idea is that we add the necessary APIs to UDisks2 (and libblockdev), but this here is a quick and dirty skeleton that should inform the API design.

There is a hidden switch that needs to be explicitly flipped before this appears in the UI.  So I think we can actually consider merging and releasing this.

The hidden switch:  Type the following into the browser console and reload the storage page.
```
window.localStorage["clevis-feature"] = "on"
```

### Setting up tang

If you want to play with this, you need to set up a `tang` server.   This should work on Fedora 27:
```
# dnf install tang
# cat <<EOF >/etc/systemd/system/tangd.socket.d/port.conf 
[Socket]
ListenStream=
ListenStream=88
EOF
# systemctl enable --now tangd.socket
```

This puts the tangs server on port 88 instead of the default 80.  Any other port is fine as well.  (If port 80 is fine, you don't even need to put the file into with /etc/systemd.)

Then install the `clevis-luks` package on the system where you want to use this.

### Tasks

- [x] Add that hidden switch
- [x] Button for refreshing a key
- [x] Simple test to demonstrate feasibility

### CODE REVIEW

**Comment by mvollmer** [comment]
@garrett, fyi.

**Comment by garrett** [comment]
Overall, I like what you've done and think you're definitely on the right track.

There are a couple of places that stick out as odd to me.

I'll write them in additional comments.

**Comment by garrett** [comment]
## Key dialog wording

I find the wording a bit odd on this screenshot (extracted from the video):

nbde-thumbnails-19

I'm not 100% sure exactly what's going on, but from my understanding it's basically asking someone to check the key with what they expect. One question that might impact this is: Where should they find the information to check to make sure these are the right keys?

Here's a quick text-based mockup of what it could look like with multiple keys:

> ### Verify Signing Keys
> 
> The server contains these keys:
> - `S354DA5phmOJmPFVQwQ_Is39Y4`
> - `4e5c3a51bdba1c1f9f36eb8e6d`
>
> [ Cancel ] **[ Trust Keys ]**

I've made the key monospaced, as that should help to distinguish problematic characters, such as l, I, and |; 0 and O; 5 and S; etc. Also, it makes mulitple keys the same length.

When there's only 1 key, it probably makes sense to change the language to:
1. Verify Signing Key
2. The server contains this key:
3. Trust Key

Trust would import and trust the key. "Trust" is probably the right verbage (versus "Import"), as it's really the primary action you'd be doing, right?

**Comment by garrett** [comment]
## Network keys

The "Network keys" entry here (right above the mouse cursor) looks oddly formatted.

nbde-thumbnails-24

Initial quick thoughts:
1. make this a (non-styled) table, so keys will line up when there is more than one
2. make the key itself monospaced (for reasons listed above)

**Comment by garrett** [comment]
Also, where you narrate and say there should be a dialog, as it's a destructive action — you're right. I agree. :wink:

**Comment by mvollmer** [comment]
> I'm not 100% sure exactly what's going on, but from my understanding it's basically asking someone to check the key with what they expect. One question that might impact this is: Where should they find the information to check to make sure these are the right keys?

Yes, we should be able to give much more help here.  I found this in `man clevis-encrypt-tang`:

```
Obtaining the thumbprint of a trusted signing key is easy.  If you have access to the Tang server's database directory, simply do:

              $ jose jwk thp -i $DBDIR/$SIG.jwk
```
I think they mean this non-ironically.  :-)

My plan is to offer this method from Cockpit:

```
Tang  can  also  perform entirely offline encryption if you pre-share the server advertisement.  You can fetch the advertisement with a simple command (just be careful your network
       isn't compromised!):

              $ curl -f $URL/adv > adv.jws

       Once you have the advertisement file, just provide it:

              $ clevis encrypt tang '{"url":...,"adv":"adv.jws"}' < PT > JWE
```

The `just be careful` part would be done by using SSH.  We can bring up a dialog similar to the troubleshooting dialog when connecting to a machine: The user would be able to type in login/password etc.  I am not familiar with that part of Cockpit, so I don't know if I can easily add this dialog to this playground already.

**Comment by mvollmer** [comment]
@garret, I was thinking about how to display the keys.  I think it would be nice to show

 - address of server
 - creation date of the key
 - any arbitrary comment attached to the key by the creator
 - algorithm and size of the key
 - thumbprint

We don't get date and comment from tang, I think.  I can ask to have them added.  Wdyt?

**Comment by garrett** [comment]
Oh! Excellent! So Cockpit can check and verify the keys itself? 

If so, it's probably still good to have a message about how it was verified automatically. And then we'd want error dialogs, like if it tried to verify and it failed due to a mismatch or some server error happened.

**Comment by Garret** [comment]
Is it that difficult to avoid writing my username? Every time I receive a notification of your project, I hope you understand that it is very annoying.

**Comment by mvollmer** [comment]
> Verify Signing Keys

Now that I look at this closer, I am a bit surprised that we are talking about "signing keys" here.  The operation we are doing is encryption, not signing.  But `clevis encrypt tang` itself talks about signing keys, so I guess I have to learn more about this.  Also, it's not super clear to me in what case we get more than one signing key, and what that means.

**Comment by mvollmer** [comment]
> Is it that difficult to avoid writing my username?

Yes, every once in a while GitHub autocompletes "@garr" to "@garret" instead of "@garrett", and I don't always notice.  This time I noticed within 5 seconds, but GitHub is faster than that, unfortunately.

Can you block notifications from me, as if I were a bad person?

**Comment by Garret** [comment]
It is just you and @martinpitt every time. I have found an option to block your notifications. I hope it will help.

**Comment by mvollmer** [comment]
>> Verify Signing Keys
>
> Now that I look at this closer, I am a bit surprised that we are talking about "signing keys" here. 

Okay, it goes like this:

- The server sends an "advertisement".  This advertisement contains a set of keys for various purposes.

- The advertisement is signed by one or more keys.  The (public halves of the) signing keys are included in the advertisement itself, and all of the included keys have signed the advertisement.

- If you encrypt something with the advertisement (such as the disk), then clevis will choose a suitable key for that from the advertisement.  That is not going to be a signing key, but a "deriveKey" key.

- Normally, the advertisement includes exactly one signing key, and exactly one "deriveKey" key.

- The magic numbers we show are the thumbprints of the signing keys in the advertisement.  All of them have signed the advertisement.

Thus, those magic numbers are not suitable to identify the key that has been used to encrypt the disk.  We can find that key afterwards, but we would have to guess beforehand.  (It's the first "deriveKey" key, but maybe that will change.)


So, I have removed the magic numbers from the "Network keys" list, and changed the text of the dialog to talk about signatures.

**Comment by mvollmer** [comment]
> I hope it will help.

@Garret, let's find out! :-)

**Comment by garrett** [comment]
- Are the keys only used for this purpose?
- How often is there more than one key?

If the keys could be used for something else, and there is commonly more than one key, we _might_ want to consider using checkboxes (checked by default) in front of every key, so if someone _didn't_ want to use a key for this, then they could unselect it.

However, if the keys are used for this purpose only, then is it acceptable to always import and verify all keys?

**Comment by mvollmer** [comment]
> - Are the keys only used for this purpose?

Potentially, no.

> - How often is there more than one key?

I can't say, unfortunately.  I should ask, once I have my own picture more in focus.

> If the keys could be used for something else, and there is commonly more than one key, we _might_ want to consider using checkboxes (checked by default) in front of every key, so if someone _didn't_ want to use a key for this, then they could unselect it.

Yeah.  Ultimately, only one key is used, and we could let the user select it.  The current clevis command line tools don't allow this, so maybe it is not important, though.

> However, if the keys are used for this purpose only, then is it acceptable to always import and verify all keys?

What I am thinking now (and I might think differently in five minutes), is that we ignore all the structure that is inside the advertisement, and treat the whole advertisement blob as "the key".

We show a simple checksum of the whole blob to the user, and ask them to verify that checksum on the server.  (That checksum is not the same as we show currently, but who cares.) A metric ton of complexity would just go away.

**Comment by mvollmer** [comment]
> I was thinking about how to display the keys. I think it would be nice to show

Let's scratch all of that for now. If we ever get to the point where we let people select a key from the advertisement, we can worry about this then.

Also, JWK has the "kid" field defined, which we can use to carry the comment.

**Comment by mvollmer** [comment]
> Oh! Excellent! So Cockpit can check and verify the keys itself?

In some situations.  Cockpit can connect to other machines over SSH.  This is what happens when you add a machine to the dashboard, for example.  So if the key server is already on your dashboard and Cockpit has the necessary credentils for logging in, Cockpit can just reach over via SSH and get the key in a secure manner.

If Cockpit doesn't have the necessary credentials, we can ask the user, just like the troubleshooting dialog on the dashboard.  This might be messy from a UX flow point of view.

Here is how the finger print dialog looks now:

verify-fp

What about putting another button in there that say essentially "Try running that command"?

**Comment by mvollmer** [comment]
@garrett, intermediate demo: https://www.youtube.com/watch?v=FZCE3nnkFLY

The audio cuts out at some point, maybe my microphone is broken as well...

One improvement I want to make: if the old and new advertisement contain at least one common signing key, we can trust the new advertisement implicitly.

I'll do this and then take another demo, maybe, with working audio hopefully.

**Comment by mvollmer** [comment]
@garrett, another demo: https://youtu.be/qPl5tl0QPNk

Right now, I am quite happy with that, from a "covering all the cases" point of view.  Of course, the result of checking the key can be shown much more eloquently, and we might do a spontaneous check in the background even.

One idea is the have a global button that checks all the keys, and the results pop up as little icons next to the keys.

**Comment by garrett** [comment]
I'm a bit concerned with the manual verification for the following reasons:

1. It adds to a wall of text that makes it more likely people may skip.
   - A possible work-around is to hide it in an expander — but only if it's optional
   - Another approach is to reduce the amount of text (which is usually a good idea anyway)
2. Running the command requires signing into the server or switching to the terminal page (which is currently problematic with copy and pasting text).
3. The suggested command is pretty complex, so if someone is manually retyping, there's a strong chance there will be a typo.
4. The instructions include the usage of a `jose` command, which would need to be installed.
   1. If `jose` command isn't installed, it would produce an error.
   2. It requires installing additional software on the server.
   3. Which server? (I guess the server that Cockpit is running on? ...Or the remote server that Cockpit is connecting to?)
5. Why is this an unstrusted network connection?

**Comment by garrett** [comment]
In the case the key is fetched over an untrusted network connection, would it be possible for Cockpit to run a similar command in the background to verify the signature and only present a dialog if the signatures do not match?

**Comment by martinpitt** [comment]
@garrett: Can't pull yourself out of the swamp - comparing the fingerprint is the physical grounding in this case, which every kind of software security requires. *If* cockpit can make a trustworthy SSH connection to the server, then we don't need to ask for/show the Clevis fingerprint at all. But there are certainly a lot of cases where cockpit isn't installed on the Clevis server, or the admin of the local machine doesn't have access to it. In fact, if I were the administrator of such a Clevis server, I'd lock that down as much as possible, as it's a very big "Single Point of Data Leak Failure".

**Comment by mvollmer** [comment]
> I'm a bit concerned with the manual verification for the following reasons:

Yes, it's not nice, but it needs to be done sometimes with the way tang currently works.

This manual verification here is about the same thing as verifying the host key of a machine when connecting to it via SSH for the first time.  SSH is secure, but on the first connection, you want to be sure that you are actually connecting to the right machine, and not a machine that some attacker is diverting your traffic to.

In the case of SSH, the connection very often goes across the general Internet, so the risk is very real.
With NBDE, the connection probably happens inside a data center or via a VPN, so I would guess that there is a good chance that the connection can actually be trusted, even if it isn't encrypted and signed.

However, the point of tang is that it _can_ work securely even over an untrusted network connection, just like SSH, once everything has been setup.

> The suggested command is pretty complex, [...]

Totally.  I think tang itself should come with a utility that spits out the necessary text.

> Why is this an unstrusted network connection?

Because tang only talks HTTP, and not HTTPS.  Of course, HTTPS isn't a trivial win, since one still needs to manage certificates, but I still think it would make sense to at least offer the option.


So, my current idea is this: In the first dialog, where you specify the URL and passphrase, we add radio buttons where the user chooses how to trust the key.  The options would be

 1) I trust the network connection to the server
 2) I will manually verify the signatures on the key
 3) I want to log into the server over a secure SSH connection and retrieve the key

When selecting 3), login, password, etc boxes appear.  

Then we you press "Add", Cockpit would do the right things:

 1) Just get the key and use it
 2) Get the key and show the current second dialog with the magic jose spell
 3) Get the key via SSH

For 3) I am actually currently thinking that I try spawning the regular ssh command, instead of connecting via cockpit-ws to cockpit-bridge in the key server.  Because there might be no cockpit-bridge on the key server.

**Comment by martinpitt** [comment]
Personally I'd prefer if the decision between 1, 2, and 3 could be postponed until after  an opportunistic ssh attempt:

 * If it succeeds, there's no reason to ask about anything else
 * If it fails because
   - of missing credentials: then retrieve the key, show it, and explain about the "trusted network" vs. offer ssh connection 
   - the tang server doesn't even have ssh: retrieve the key, show it, and explain about "trusted network" vs. ascertaining  key validity by physical means

That would be easier and avoid impossible or less practical options upfront?

**Comment by mvollmer** [comment]
> Personally I'd prefer if the decision between 1, 2, and 3 could be postponed until after an opportunistic ssh attempt:

I am afraid this would be too magical, noo?  People might now know why it "just works" this time, but not that other time.  But we can explain this in the dialogs.

After talking with @tigert here (thanks!), I don't think I want to pull in the full complexity of connecting via SSH here.  Having to verify the hostkey, correcting the port, finding out that there isn't any cockpit-bridge on the other side is not really an improvement, I think.

I am in favor of scoping SSH out at this point.

**Comment by martinpitt** [comment]
> I am in favor of scoping SSH out at this point.

Agreed. It makes things a whole lot more complex, and is not really necessary for the MVP. This can become a future enhancement.

**Comment by martinpitt** [comment]
In other words, always show the host key and ask the user to physically validate the key or trust her network? I. e. the dialog woud not have any choices (apart from  "Accept" and "Cancel"), just this explanation?

**Comment by garrett** [comment]
I iterated through a couple of text based mockups of what it could look like. I'm suggesting the last one, but including the others as well, to show how I arrived at it (and also in case the last one doesn't work well enough).

---

# Iteration 1: Using an expander in the dialog

Collapsed (default):

> ### Key Verification
> Verify the key using its fingerprint:
>
> `f50dd8ldf7ccb1270afd533e46120523368e2474`
>
> [▸ How do I check the fingerprint?](#expander)
>
> [Cancel] **[Trust Key]**

Expanded:

> ### Key Verification
> Verify the key using its fingerprint:
>
> `f50dd8ldf7ccb1270afd533e46120523368e2474`
>
> [▾ How do I check the fingerprint?](#expander)
>
> Use SSH to connect to 192.168.100.1 and run the following command:
> `curl -s localhost:88/adv | sha1sum`
>
> [Cancel] **[Trust Key]**

---

# Iteration 2: Back to all-in-one

Based on the modified text from above, I guess it might be worth simplifying back into a single dialog:

> ### Verify Key
> The key fingerprint is `f50dd8ldf7ccb1270afd533e46120523368e2474`
>
> To verify the key, use SSH to connect to 192.168.100.1 and run the following command:
> `curl -s localhost:88/adv | sha1sum`
>
> [Cancel] **[Trust Key]**

# Iteration 3: Even simpler

> ### Verify Key
>  Manually verify the key on the server:
> `ssh 192.168.100.1 "curl -s localhost:88/adv | sha1sum"`
>
> The fingerprint should match `f50dd8ldf7ccb1270afd533e46120523368e2474`.
>
> [Cancel] **[Trust Key]**

- we know the server address (as it's asked for in the previous dialog)
- this provides a quick 1-liner way to check the keys
- the quotes run both `curl` and `sha1sum` on the server; dropping the quotes only runs `curl` on the server and `sha1sum` locally — will it be more likely that `sha1sum` exists on the server?
- it suggests checking the fingerprint, but doesn't demand it
- if the network is trusted, people could just accept it without checking
- I previously had "To manually verify the key on the server, run:" — this would be more of a suggestion than a command. Which is better?

One issue is that the SSH port of the server is unknown. In most cases, it will be 22 and when it differs, the admin will probably know and adjust the command accordingly.

**Comment by mvollmer** [comment]
> Manually verify the key on the server:

I like variant 3!  The exact command will hopefully be simpler, something like `ssh 192.168.100.1 tang-signatures 88`.

I had a crazy idea: What if we embed a terminal right in the dialog and prefill it somehow with the command?  People can then run it right away and deal with all the complexity, such as verifying a hostkey, changing the port, or user, or whatever else is necessary.

**Comment by garrett** [comment]
If we can run it from the dialog as-is, why not just run it before the dialog and check the results? If any of the assumptions cause it to not work, then it won't work in the terminal in the dialog either, right?

**Comment by mvollmer** [comment]
The dialog now looks like this:

verify-key

The tangd-show-signatures command doesn't exist yet, but hopefully soon: https://github.com/latchset/tang/pull/21

**Comment by mvollmer** [comment]
@garret, this is great! :-) We don't even need to tell the user what the text is that they are checking, which would be the hardest part.  Excellent!

**Comment by martinpitt** [comment]
@mvollmer: Nice! What is the magic "88" there?

**Comment by mvollmer** [comment]
> What is the magic "88" there?

That is the port that the tang server is listening on.  It comes from the URL that people have typed, so they would recognize it, I guess.

**Comment by garrett** [comment]
1. I was thinking the command would be a block and the fingerprint would be inline (and wrap all together to the next line if needed), so like the opposite of what you have.
2. Make sure you spell my name correctly. :wink:

**Comment by mvollmer** [comment]
> Make sure you spell my name correctly. wink

Oh, for effs sake! :-/

**Comment by mvollmer** [comment]
I think this is as good as it gets for now.

Next steps could be to adding "clevis luks unbind" and "clevis luks rebind" and start wrapping this all in libblockdev / UDisks2.

**Comment by mvollmer** [comment]
> Next steps could be to add "clevis luks unbind"

clevis luks unbind already exists in master, nice.

**Comment by martinpitt** [comment]
I tried to test this locally. I started a tang server (from our fedora 27 image), and installed clevis locally (this was missing from the description, I added that now), and a console.log in `enable_clevis_features()` confirms it's successful. And yet I don't see the "network keys" bit:

clevis

File: pkg/storaged/clevis-dialogs.jsx [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+                      $('<div>').append(
+                          extra ? $('<p>').text(extra) : null,
+                          $('<p>').append(
+                              $('<span>').text(_("Manually verify the key on the server: ")),
+                              $('<pre>').text(cmd)),
+                          $('<p>').append(
+                              $('<span>').text(_("The output should match this text: ")),
+                              (info.sigkeys.length == 1 ?
+                               $('<samp>').text(info.sigkeys[0]) :
+                               $('<pre>').text(info.sigkeys.join("\n"))))
```
Comment (martinpitt) [inline pkg/storaged/clevis-dialogs.jsx]: It seems a bit odd to format one or multiple keys differently. Would be simpler to always call `join()` (that DTRT for one-element arrays) and use `white-space: pre-wrap`?

File: pkg/storaged/client.js [hunk 2]
```diff
@@ -372,44 +373,65 @@
...
+        function enable_vdo_features() {
+            return cockpit.spawn([ "which", "vdo" ], { err: "ignore" }).then(
+                function () {
+                    client.features.vdo = true;
+                    client.vdo_overlay.start();
+                    return cockpit.resolve();
+                },
+                function () {
+                    return cockpit.resolve();
+                });
```
Comment (martinpitt) [inline pkg/storaged/client.js]: No need to wrap a function around this, you can just directly give `cockpit.resolve` as second argument (same below).

File: pkg/storaged/client.js [hunk 3]
```diff
@@ -769,6 +790,94 @@
...
+            remove: remove,
+            get_adv: get_adv,
+            add: add,
+            check_key: check_key,
+            replace: replace,
+            unlock: unlock
+        };
+
+        function spawn_tool(args) {
+            return cockpit.spawn([ "python3", "--", "-" ].concat(args), { superuser: "try", err: "message" })
```
Comment (martinpitt) [inline pkg/storaged/client.js]: Deliberately not using `python.spawn()` here? Any chance that this will get backported to RHEL 7? Or is this just a temporary shim until clevis/tang get more API (the comment in the py helper suggest that)?

File: pkg/storaged/client.js [hunk 4]
```diff
@@ -769,6 +790,94 @@
...
+
+        function some(array, func) {
+            var i;
+            for (i = 0; i < array.length; i++) {
+                var val = func(array[i]);
+                if (val)
+                    return val;
+            }
+            return null;
+        }
```
Comment (martinpitt) [inline pkg/storaged/client.js]: This looks like `array.find(val => func(val))`?

File: pkg/storaged/cockpit-clevis-tool.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    res = { }
+    for dev in subprocess.check_output([ "lsblk", "-pln", "-o", "NAME" ]).splitlines():
+        i = info(dev)
+        if i:
+            res[dev.decode()] = i
+    return res
+
+def monitor():
+    mon = subprocess.Popen([ "stdbuf", "-o", "L", "udevadm", "monitor", "-u", "-s", "block"],
+                           bufsize=1, stdout=subprocess.PIPE)
```
Comment (martinpitt) [inline pkg/storaged/cockpit-clevis-tool.py]: Urgh, I hope that will go away -- udisks devices or systemd device units (both D-Bus accessible through signals) don't have the information you need? This doesn't even use `--env` to show the properties, so it looks like this literally just watches for new/removed block devices?

File: pkg/storaged/cockpit-clevis-tool.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    subprocess.run([ "luksmeta", "wipe", "-d", dev, "-s", slot, "-f" ])
+    subprocess.run([ "luksmeta", "save", "-d", dev, "-s", slot, "-u", "cb6e8904-81ff-40da-a84a-07ab9ab5715e" ],
+                   check = True, input = new_jwe)
+
+def unlock(dev):
+    # XXX - clevis-luks-unlock always exits 1, so we check whether the
+    # expected device exists afterwards.
+    clear_dev = b"luks-" + subprocess.check_output([ "cryptsetup", "luksUUID", dev ]).strip()
+    subprocess.run([ "clevis", "luks", "unlock", "-d", dev, "-n", clear_dev ])
+    sys.exit(0 if os.path.exists(b"/dev/mapper/" + clear_dev) else 1)
```
Comment (martinpitt) [inline pkg/storaged/cockpit-clevis-tool.py]: This is guaranteed to be synchronous, i. e. `clevis luks unlock` always waits until the cleartext device appears? Or might this need a waiting loop?

File: test/verify/check-storage-luks [hunk 7]
```diff
@@ -123,5 +123,80 @@ class TestStorage(StorageCase):
...
+        self.content_tab_action(1, 1, "Add")
+        self.dialog_wait_open()
+        self.dialog_set_val("url", "127.0.0.1")
+        self.dialog_set_val("passphrase", "vainu-reku-toma-rolle-kaja")
+        self.dialog_apply()
+        b.wait_in_text("#dialog", "The output should match this text")
+        b.wait_in_text("#dialog", sig_thp)
+        self.dialog_apply()
+        self.dialog_wait_close()
+        self.content_tab_wait_in_info(1, 1, "Network keys", "127.0.0.1")
```
Comment (martinpitt) [inline test/verify/check-storage-luks]: THis doesn't test most of key removal/obsoletion etc. yet, but could you at least lock the device again, unlock it, and make sure this succeeds without typing a passphrase? So that we cover the most basic use case at least.

**Comment by mvollmer** [comment]
> And yet I don't see the "network keys" bit:

@martinpitt, It should be in the "Encryption" tab.

File: pkg/storaged/clevis-dialogs.jsx [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+                  Body: (
+                      $('<div>').append(
+                          extra ? $('<p>').text(extra) : null,
+                          $('<p>').append(
+                              $('<span>').text(_("Manually verify the key on the server: ")),
+                              $('<pre>').text(cmd)),
+                          $('<p>').append(
+                              $('<span>').text(_("The output should match this text: ")),
+                              (info.sigkeys.length == 1 ?
+                               $('<samp>').text(info.sigkeys[0]) :
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/storaged/clevis-dialogs.jsx]: Hmm, it should be monospace, and <samp> gives us that.  <code> also works, but Patternfly gives it a blue background. @garrett?

File: pkg/storaged/clevis-dialogs.jsx (hunk 1 above)
Comment (mvollmer) [reply pkg/storaged/clevis-dialogs.jsx]: `samp` is inline, while `pre` is a block. @garrett?

File: pkg/storaged/client.js (hunk 3 above)
Comment (mvollmer) [reply pkg/storaged/client.js]: It's all temporary.  The plan is to add enough API to UDisks2 for this.

File: pkg/storaged/cockpit-clevis-tool.py (hunk 5 above)
Comment (mvollmer) [reply pkg/storaged/cockpit-clevis-tool.py]: Yes, this will go away.  This is just to get change notifications when luksmeta has written something new to a disk.  When UDisks2 does it, it will fit nicely into its existing udev event loop.

File: pkg/storaged/cockpit-clevis-tool.py (hunk 6 above)
Comment (mvollmer) [reply pkg/storaged/cockpit-clevis-tool.py]: It's synchronous, thank god.

File: test/verify/check-storage-luks (hunk 7 above)
Comment (mvollmer) [reply test/verify/check-storage-luks]: Yes, that would be worthwhile indeed.

File: pkg/storaged/clevis-dialogs.jsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/storaged/clevis-dialogs.jsx]: Oh, good. Sorry, it really looked like a typo, and I've never encountered `<samp>`.

**Comment by martinpitt** [comment]
> It should be in the "Encryption" tab.

Oh, stupid me, thanks. Now I get further. Key validation on the server fails with `tang-show-keys: command not found` - I suppose this is a recent upstream addition which just didn't land in Fedora 27 yet? So I just said "trust this key". But now the "Encryption" tab didn't change, and when I press "Unlock" it still wants a passphrase. I thought the intention was to show the tang server IP at least, and the fact that there is a network key?

**Comment by martinpitt** [comment]
Oh, of course I saw the error right after pressing "send"...

    FileNotFoundError: [Errno 2] No such file or directory: 'luksmeta': 'luksmeta'

I adjusted the description accordingly. So I installed that, but same behaviour. Now I get this error message:
```
/!\ 
Usage: clevis COMMAND [OPTIONS]

  clevis decrypt      Decrypts using the policy defined at encryption time
  clevis encrypt http Encrypts using a REST HTTP escrow server policy
  clevis encrypt sss  Encrypts using a Shamir's Secret Sharing policy
  clevis encrypt tang Encrypts using a Tang binding server policy
```

**Comment by mvollmer** [comment]
> tang-show-keys: command not found - I suppose this is a recent upstream addition which just didn't land in Fedora 27 yet? 

Correct.

**Comment by mvollmer** [comment]
>  But now the "Encryption" tab didn't change,

Didn't you get an error in the dialog?

In any case, I think installing `clevis-luks` should do the trick.

**Comment by mvollmer** [comment]
> Didn't you get an error in the dialog?

Indeed, there is no error when clevis-luks isn't installed.  Needs fixing.

**Comment by mvollmer** [comment]
> Indeed, there is no error when clevis-luks isn't installed. Needs fixing.

https://github.com/latchset/clevis/issues/42

I don't think we need to work around this.

File: pkg/storaged/clevis-dialogs.jsx [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import cockpit from "cockpit";
+import dialog from "./dialog";
+import $ from "jquery";
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/storaged/clevis-dialogs.jsx]: I try to use React!

File: pkg/storaged/clevis-dialogs.jsx [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+var _ = cockpit.gettext;
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/storaged/clevis-dialogs.jsx]: Fixed.

File: pkg/storaged/clevis-dialogs.jsx [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+function add_adv(client, block, url, info, passphrase) {
+    verify_adv(url, info,
+               _("Verify Key"),
+               null,
+               _("Trust Key"),
+               function () {
+                   return client.clevis_overlay.add(block, url, info.adv, passphrase)
+                                .fail(function (err) {
+                                    console.log("FAIL", err);
+                                });
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/storaged/clevis-dialogs.jsx]: That was debugging code.

File: pkg/storaged/clevis-dialogs.jsx (hunk 1 above)
Comment (mvollmer) [reply pkg/storaged/clevis-dialogs.jsx]: With one signing key:
screenshot from 2018-03-26 14-23-20

With two:
screenshot from 2018-03-26 14-24-20

File: pkg/storaged/client.js (hunk 2 above)
Comment (mvollmer) [reply pkg/storaged/client.js]: But then `this` will be different by the time `cockpit.resolve` is actually called, no?  It doesn't matter in this case, but that difference and the fact that it breaks symmetry between the two cases makes me prefer the long form actually.

File: pkg/storaged/client.js (hunk 4 above)
Comment (mvollmer) [reply pkg/storaged/client.js]: You mean `array.find(func)`? :-)  `some` returns the first non-false return value of `func`, while `find` returns a element of the array itself.

File: pkg/storaged/clevis-dialogs.jsx (hunk 1 above)
Comment (garrett) [reply pkg/storaged/clevis-dialogs.jsx]: Sorry for the delay; I wrote this earlier (more or less) and GitHub didn't save this comment:

Yep:
- `pre` is block by default
- `samp` is inline by default (for computer output)

Related:
- `kbd` is inline for typed input to the computer
- `code` is for code snippets

Notes:

- `samp`, `kbd`, and `code` are all inline and often should be wrapped in a `pre`
  - `pre` just means prefomatted text, whereas the rest of the tags mentioned here all provide semantic meaning (and might be wrapped in `pre`, but not necessarily)
- `kbd` and `samp` can be nested in each other
- `kbd` inside of `kbd` signifies a key to type; example: `<kbd><kbd>Ctrl</kbd> + <kbd>z</kbd></kbd>`

You can think of `samp` and `kbd` as using single backticks inline in Markdown, and `pre` as the triple-backticks for preformatted samples (and `<pre><code></code></pre>` for code samples).

You might want to also use `kbd` for the string that should be typed.

---

**TL;DR**: Correct. :+1: Consider using `kbd` too.

File: test/verify/check-storage-luks (hunk 7 above)
Comment (mvollmer) [reply test/verify/check-storage-luks]: Now the test cover a lot more, including key rotation.

File: pkg/storaged/clevis-dialogs.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/storaged/clevis-dialogs.jsx]: Done!

File: pkg/storaged/clevis-dialogs.jsx (hunk 1 above)
Comment (mvollmer) [reply pkg/storaged/clevis-dialogs.jsx]: @garrett, whoa, thanks!  Could you make the call here?  Is the current code okay, or should I change something?

Comment (garrett) [reply pkg/storaged/clevis-dialogs.jsx]: > Consider using `kbd`

... for text that should be typed.

```html
<pre><kbd>ssh ...</kbd></pre>

<pre><samp>mGSngj...</samp></pre>
```

The `<pre>` is there because these are blocks. If you had the text inline, you wouldn't have the `pre`, but would have the `kbd` & `samp` still.

If you have time to do a quick fix in this PR, that would be great. Pushing this to another PR afterward is fine too, as long as it's fixed.

Comment (mvollmer) [reply pkg/storaged/clevis-dialogs.jsx]: Looks like this now:
screenshot from 2018-03-27 11-59-32

Comment (garrett) [reply pkg/storaged/clevis-dialogs.jsx]: Oh. Well, I guess it's styled up in PatternFly or within Cockpit CSS.

Drop the `<kbd>` for now, and I guess we should look into fixing it at a more general level first.

Comment (mvollmer) [reply pkg/storaged/clevis-dialogs.jsx]: The gray-on-black colors come from Patternfly.

Comment (mvollmer) [reply pkg/storaged/clevis-dialogs.jsx]: Done.

Comment (garrett) [reply pkg/storaged/clevis-dialogs.jsx]: Opened issues @ https://github.com/cockpit-project/cockpit/issues/8904 & https://github.com/patternfly/patternfly/issues/1002
